src/core/symbols/html_symbol_report.cpp

```diff
@@ -0,0 +1,285 @@
+ */
+
+#include "html_symbol_report.h"
+
+#include <QBuffer>
+#include <QCoreApplication>
+#include <QImage>
+#include <QImageWriter>
+#include <QLocale>
+#include <QString>
+#include <QTextDocumentFragment>
+
+#include "core/map.h"
+#include "core/map_color.h"
+#include "core/symbols/symbol.h"
+
+namespace OpenOrienteering {
+
+/**
+ * Generates HTML reports for colors and symbols in a Map.
+ * 
+ * The public interface of this class is the free-standing function
+ * makeHTMLSymbolReport(const Map& map).
+ */
+class HTMLSymbolReportGenerator
+{
+private:
+	/**
+	 * Efficiently produce PNG data from QImage.
+	 * 
+	 * An object of this class holds a buffer which is reused in multiple calls
+	 * to write(). So the number of actual memory allocations can kept small.
+	 */
+	class PNGImageWriter
+	{
+		QByteArray png {"PNG"};
+		QByteArray png_data;
+		
+	public:
+		/**
+		 * Creates PNG data from an image.
+		 * 
+		 * @return A reference to a shared buffer. The buffer is invalidated
+		 *         when the function is called again (for the same writer).
+		 */
+		const QByteArray& write(const QImage& image)
+		{
+			png_data.clear();
+			QBuffer buffer(&png_data);
+			QImageWriter(&buffer, png).write(image);
+			return png_data;
+		}
+	};
+	
+	enum class ColorRowType {
+		Basic,    ///< A color row with number, icon and name
+		Extended  ///< A color row with all details
+	};
+	
+	const Map& map;
+	QImage icon_image;
+	QLocale locale;
+	PNGImageWriter image_writer;
+	
+	QString imgForColor(const QColor& c, const QString& alt)
+	{
+		icon_image.fill(c);
+		return QString::fromLatin1("<img alt=\"%1\" src=\"data:image/png;base64,").arg(alt)
+		       + QString::fromLatin1(image_writer.write(icon_image).toBase64())
+		       + QString::fromLatin1("\">");
+	}
+	
+	QString makeColorRow(const MapColor& c, ColorRowType row_type)
+	{
+		icon_image = QImage(16, 16, QImage::Format_ARGB32);
+		
+		auto details = QString {};
+		if (row_type == ColorRowType::Extended)
+		{
+			auto cmyk = c.getCmyk();
+			details = QString::fromLatin1(
+			              "<td>%1</td>" // c
+			              "<td>%2</td>" // m
+			              "<td>%3</td>" // y
+			              "<td>%4</td>" // k
+			              "<td style=\"font-family:monospace\">%5</td>" // rgb
+			              "<td>%6</td>" // spot color
+			              "<td>%7</td>" // knockout
+			              ).arg(
+			                  locale.toString(100*cmyk.c, 'g', 3),
+			                  locale.toString(100*cmyk.m, 'g', 3),
+			                  locale.toString(100*cmyk.y, 'g', 3),
+			                  locale.toString(100*cmyk.k, 'g', 3),
+			                  QColor(c.getRgb()).name(),
+			                  QString{c.getSpotColorName()}.replace(QString::fromLatin1(", "), QString::fromLatin1(",<br>")),
+			                  c.getKnockout() ? QString::fromLatin1("[X]") : QString{} );
+		}
+		
+		auto name = QTextDocumentFragment::fromHtml(c.getName()).toPlainText();
```

Comment:
Hm, this is not enough to transform random text into safe text for the `alt` attribute. We must be concerned about the special characters `"`, `<`, `>`, `&`. But I don't have a simple function at hand.